File: frontend/src/lib/clipAnim.js
// Pose del clip en un instante: keyframes (snapshots) o pistas `anim` antiguas.
// `t` es tiempo LOCAL del clip (0 = inicio de la barra).

import {
  clipPropsAt, keyframesEnabled, staticProps,
} from './clipKeyframes.js'

export const ANIM_PROPS = ['x', 'y', 'scale', 'rotation', 'opacity']

function num(v, d) {
  const n = Number(v)
  return Number.isFinite(n) ? n : d
}

export function normalizeTrack(kfs) {
  return (kfs || [])
    .map((k) => ({
      t: num(k?.t, NaN),
      v: num(k?.v, NaN),
      ease: k?.ease === 'direct' ? 'direct' : 'smooth',
    }))
    .filter((k) => Number.isFinite(k.t) && Number.isFinite(k.v))
    .sort((a, b) => a.t - b.t)
}

/** Valor de una pista en `t`. `ease` del keyframe de llegada (como pan_mode del encuadre). */
export function interpTrack(kfs, t, fallback) {
  const s = normalizeTrack(kfs)
  const fb = num(fallback, 0)
  if (!s.length) return fb
  const time = num(t, 0)
  if (time <= s[0].t) return s[0].v
  const last = s[s.length - 1]
  if (time >= last.t) return last.v
  for (let i = 0; i < s.length - 1; i++) {
    const a = s[i], b = s[i + 1]
    if (time >= a.t && time <= b.t) {
      if (b.ease === 'direct') return a.v
      const f = (time - a.t) / ((b.t - a.t) || 1)
      return a.v + (b.v - a.v) * f
    }
  }
  return last.v
}

export function staticPose(clip) {
  const p = staticProps(clip)
  return {
    x: p.x, y: p.y, scale: p.scale, rotation: p.rotation, opacity: p.opacity,
  }
}

function legacyAnimPose(clip, localT, base) {
  const anim = clip?.anim && typeof clip.anim === 'object' && !Array.isArray(clip.anim) ? clip.anim : {}
  return {
    x: interpTrack(anim.x, localT, base.x),
    y: interpTrack(anim.y, localT, base.y),
    scale: interpTrack(anim.scale, localT, base.scale),
    rotation: interpTrack(anim.rotation, localT, base.rotation),
    opacity: interpTrack(anim.opacity, localT, base.opacity),
  }
}

export function clipPose(clip, localT, srcTime) {
  const p = clipPropsAt(clip, localT, srcTime)
  if (keyframesEnabled(clip)) {
    return {
      x: p.x, y: p.y, scale: p.scale, rotation: p.rotation, opacity: p.opacity,
      cx: p.cx, cy: p.cy, zoom: p.zoom,
    }
  }
  const legacy = legacyAnimPose(clip, localT, p)
  return { ...legacy, cx: p.cx, cy: p.cy, zoom: p.zoom }
}

export function posedTransform(clip, localT) {
  const p = clipPose(clip, localT)
  return { x: p.x, y: p.y, scale: p.scale, rotation: p.rotation }
}

export function applyShapePose(st, pose) {
  const scale = num(pose?.scale, 1)
  return {
    ...st,
    x: pose.x,
    y: pose.y,
    rotation: pose.rotation,
    opacity: pose.opacity,
    w: num(st?.w, 0.38) * scale,
    h: num(st?.h, 0.16) * scale,
  }
}

File: backend/app/clip_anim.py
"""Animación de clip (posición / escala / rotación / opacidad).

Espejo de ``frontend/src/lib/clipAnim.js``. ``t`` es tiempo local del clip.
"""
from __future__ import annotations

from typing import Any, Optional

from .clip_keyframes import clip_props_at, keyframes_enabled, static_props


def _num(v: Any, default: float) -> float:
    try:
        n = float(v)
    except (TypeError, ValueError):
        return default
    if n != n:  # NaN
        return default
    return n


def normalize_track(kfs: Optional[list]) -> list[dict]:
    out = []
    for k in kfs or []:
        if not isinstance(k, dict):
            continue
        t = _num(k.get("t"), float("nan"))
        v = _num(k.get("v"), float("nan"))
        if t != t or v != v:
            continue
        out.append({"t": t, "v": v, "ease": "direct" if k.get("ease") == "direct" else "smooth"})
    out.sort(key=lambda k: k["t"])
    return out


def interp_track(kfs: Optional[list], t: float, fallback: float) -> float:
    s = normalize_track(kfs)
    fb = _num(fallback, 0.0)
    if not s:
        return fb
    time = _num(t, 0.0)
    if time <= s[0]["t"]:
        return s[0]["v"]
    last = s[-1]
    if time >= last["t"]:
        return last["v"]
    for i in range(len(s) - 1):
        a, b = s[i], s[i + 1]
        if a["t"] <= time <= b["t"]:
            if b["ease"] == "direct":
                return a["v"]
            span = (b["t"] - a["t"]) or 1.0
            f = (time - a["t"]) / span
            return a["v"] + (b["v"] - a["v"]) * f
    return last["v"]


def static_pose(clip: Any) -> dict:
    p = static_props(clip)
    return {k: p[k] for k in ("x", "y", "scale", "rotation", "opacity")}


def clip_pose(clip: Any, local_t: float) -> dict:
    if keyframes_enabled(clip):
        p = clip_props_at(clip, local_t)
        return {k: p[k] for k in ("x", "y", "scale", "rotation", "opacity")}
    base = static_pose(clip)
    anim = (clip.get("anim") if isinstance(clip, dict) else getattr(clip, "anim", None)) or {}
    if not isinstance(anim, dict):
        anim = {}
    return {
        "x": interp_track(anim.get("x"), local_t, base["x"]),
        "y": interp_track(anim.get("y"), local_t, base["y"]),
        "scale": interp_track(anim.get("scale"), local_t, base["scale"]),
        "rotation": interp_track(anim.get("rotation"), local_t, base["rotation"]),
        "opacity": interp_track(anim.get("opacity"), local_t, base["opacity"]),
    }
